v2/pkg/engine/datasource/graphql_datasource/graphql_subscription_client.go

```diff
@@ -387,3 +530,81 @@ func waitForAck(ctx context.Context, conn *websocket.Conn) error {
 		}
 	}
 }
+
+func (c *subscriptionClient) runEpoll(ctx context.Context) {
+	var (
+		ticker <-chan time.Time
+		done   = ctx.Done()
+	)
+	if c.epollConfig.Wait > 0 {
+		tick := time.NewTicker(time.Millisecond * 50)
+		defer tick.Stop()
+		ticker = tick.C
+	} else {
+		tick := make(chan time.Time)
+		close(tick)
+		ticker = tick
+	}
+	for {
+		connections, err := c.epoll.Wait(50)
+		if err != nil {
+			c.log.Error("epoll.Wait", abstractlogger.Error(err))
+			return
+		}
+		c.connectionsMu.Lock()
+		for _, conn := range connections {
+			id := epoller.SocketFD(conn)
+			handler, ok := c.connections[id]
+			if !ok {
+				continue
+			}
+			c.activeConnectionsMu.Lock()
+			_, active := c.activeConnections[id]
+			if !active {
+				c.activeConnections[id] = struct{}{}
+			}
+			c.activeConnectionsMu.Unlock()
+			if active {
+				continue
+			}
+			go c.handleConnection(id, handler, conn)
```

Comment:
As discussed. We should try to limit the concurrency here without service interruption.